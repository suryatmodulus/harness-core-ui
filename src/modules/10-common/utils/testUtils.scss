/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.browser {
  border: 1px solid #cccccc;
  border-radius: 8px;
  min-height: 300px;
  overflow: auto;

  .browser-header {
    background-color: #cccccc;
    padding: 8px 16px;
    display: flex;
  }

  .browser-path {
    display: block;
    border: none;
    line-height: 14px;
    font-size: 12px;
    color: #444444;
    border-radius: 9px;
    min-width: 100px;
    padding: 2px 8px;
    flex: 1 1 auto;
    border: 1px solid transparent;

    &:focus {
      border-color: rgb(30, 167, 253);
      outline: none;
    }
  }

  .browser-content {
    padding: 8px;
  }
}
