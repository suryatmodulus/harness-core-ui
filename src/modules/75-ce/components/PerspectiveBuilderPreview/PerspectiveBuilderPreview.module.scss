/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.groupByLabel {
  width: 360px;

  .labelValueContainer {
    height: 370px;
    overflow-y: auto;
  }
}

.groupBySelect {
  padding: var(--spacing-xsmall) var(--spacing-small) var(--spacing-xsmall) var(--spacing-small) !important;
  border-radius: 4px;
  max-width: 200px;
  overflow: hidden;
  margin-top: var(--spacing-xsmall) !important;
  cursor: pointer;

  &.groupBySelectLoading {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 80px;
  }
}

.previewMainContainer {
  overflow-y: auto;
}

.menuItem {
  padding: var(--spacing-small);
  &:hover {
    background: var(--primary-1) !important;
    color: var(--black) !important;
  }
  &:active {
    background: var(--primary-7) !important;
    color: var(--white) !important;
  }
}
