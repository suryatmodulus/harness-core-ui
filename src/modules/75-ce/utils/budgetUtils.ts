/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export enum AlertStatus {
  Danger,
  Warning,
  Info
}

interface CostStatusInfo {
  cost: number
  ratio: number
  percentage: number
  status: AlertStatus
}

export interface getAllBudgetCostInfoReturnType {
  actualCostStatus: CostStatusInfo
  forecastedCostStatus: CostStatusInfo
  budgetAmountStatus: CostStatusInfo
}

export const getAllBudgetCostInfo: (
  forecastedCost: number,
  actualCost: number,
  budgetAmount: number
) => getAllBudgetCostInfoReturnType = (forecastedCost, actualCost, budgetAmount) => {
  /**
   * Find out the max cost of all and it will be considered as 100%
   * and the rest of the cost percentages will be calculated based on that
   */

  const baseCost = Math.max(actualCost, forecastedCost, budgetAmount)
  const actualCostRatio = actualCost / baseCost
  const forecastCostRatio = forecastedCost / baseCost
  const budgetAmountRatio = budgetAmount / baseCost

  const actualCostPerWithBudget = (actualCost / budgetAmount) * 100
  const forecastCostPerWithBudget = (forecastedCost / budgetAmount) * 100

  const getAlertStatus: (percentage: number) => AlertStatus = percentage => {
    if (percentage > 100) {
      return AlertStatus.Danger
    }
    if (percentage > 90) {
      return AlertStatus.Warning
    }
    return AlertStatus.Info
  }

  const actualCostStatus: CostStatusInfo = {
    cost: actualCost,
    ratio: actualCostRatio,
    percentage: actualCostPerWithBudget,
    status: getAlertStatus(actualCostPerWithBudget)
  }
  const forecastedCostStatus: CostStatusInfo = {
    cost: forecastedCost,
    ratio: forecastCostRatio,
    percentage: forecastCostPerWithBudget,
    status: getAlertStatus(forecastCostPerWithBudget)
  }
  const budgetAmountStatus: CostStatusInfo = {
    cost: budgetAmount,
    ratio: budgetAmountRatio,
    percentage: 100,
    status: AlertStatus.Info
  }

  return {
    actualCostStatus,
    forecastedCostStatus,
    budgetAmountStatus
  }
}
