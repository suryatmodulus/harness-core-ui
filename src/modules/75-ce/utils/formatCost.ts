/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

interface FormatCostOptions {
  currency?: string
  locale?: string
  decimalPoints?: number
  shortFormat?: boolean
}

type FormatCost = (value: number, options?: FormatCostOptions) => string

const defaultOptions = {
  shortFormat: false,
  currency: 'USD',
  locale: 'en-us',
  decimalPoints: 2
}

const TOTAL_DIGITS_TO_SHOW = 3
const POWER_OF_TEN_IN_BILLION = 9
const POWER_OF_TEN_IN_MILLION = 6
const POWER_OF_TEN_IN_THOUSANDS = 3

type convertNumberToLocaleStringType = (
  value: number,
  locale: string,
  currency: string,
  maximumFractionDigits: number,
  minimumFractionDigits: number
) => string

const convertNumberToLocaleString: convertNumberToLocaleStringType = (
  value,
  locale,
  currency,
  maximumFractionDigits,
  minimumFractionDigits
) =>
  value.toLocaleString(locale, {
    style: 'currency',
    currency,
    maximumFractionDigits,
    minimumFractionDigits
  })

const formatCost: FormatCost = (value: number, options?: FormatCostOptions) => {
  const costOptions = { ...defaultOptions, ...options }
  if (isNaN(value)) {
    return ''
  }

  const getCostInShortFormat = (power: number, cost: number, suffix: string, numberOfDigits: number) => {
    const decimalPlacesToHave = Math.max(power + TOTAL_DIGITS_TO_SHOW - numberOfDigits, 0)
    return (
      convertNumberToLocaleString(
        cost / Math.pow(10, power),
        costOptions.locale,
        costOptions.currency,
        decimalPlacesToHave,
        decimalPlacesToHave
      ) + suffix
    )
  }

  /**
    Option shortFormat will convert the cost as mentioned below -
    2,345,456 as $2.34M
    12,345,456 as $12.3M
    123,456,789 as $123M
    and so on for thousands and billions
  */
  if (costOptions.shortFormat) {
    const numberOfDigits = value.toFixed(0).length
    if (numberOfDigits > POWER_OF_TEN_IN_BILLION) {
      return getCostInShortFormat(POWER_OF_TEN_IN_BILLION, value, 'B', numberOfDigits)
    }

    if (numberOfDigits > 6) {
      return getCostInShortFormat(POWER_OF_TEN_IN_MILLION, value, 'M', numberOfDigits)
    }

    if (numberOfDigits > 3) {
      return getCostInShortFormat(POWER_OF_TEN_IN_THOUSANDS, value, 'K', numberOfDigits)
    }
  }

  return convertNumberToLocaleString(
    value,
    costOptions.locale,
    costOptions.currency,
    costOptions.decimalPoints,
    costOptions.decimalPoints
  )
}

export default formatCost
