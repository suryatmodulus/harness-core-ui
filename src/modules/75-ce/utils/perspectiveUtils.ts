/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@wings-software/uicore'
import type Highcharts from 'highcharts'
import type { ViewRule, ViewIdCondition } from 'services/ce'
import {
  QlceViewTimeFilterOperator,
  ViewFieldIdentifier,
  QlceViewFilterWrapperInput,
  QlceViewFieldInputInput,
  QlceViewGroupByInput,
  QlceViewTimeGroupType,
  QlceViewFilterInput,
  ViewTimeRangeType,
  ViewChartType,
  QlceViewRuleInput
} from 'services/ce/services'
import { DATE_RANGE_SHORTCUTS } from './momentUtils'

const startTimeLabel = 'startTime'

export const getViewFilterForId: (viewId: string, isPreview?: boolean) => QlceViewFilterWrapperInput = (
  viewId: string,
  isPreview = false
) => {
  return { viewMetadataFilter: { viewId: viewId, isPreview: isPreview } } as QlceViewFilterWrapperInput
}

export const getTimeFilters: (from: number, to: number) => QlceViewFilterWrapperInput[] = (from, to) => {
  return [
    {
      timeFilter: {
        field: { fieldId: startTimeLabel, fieldName: startTimeLabel, identifier: ViewFieldIdentifier.Common },
        operator: QlceViewTimeFilterOperator.After,
        value: from
      }
    },
    {
      timeFilter: {
        field: { fieldId: startTimeLabel, fieldName: startTimeLabel, identifier: ViewFieldIdentifier.Common },
        operator: QlceViewTimeFilterOperator.Before,
        value: to
      }
    }
  ] as QlceViewFilterWrapperInput[]
}

export const getGroupByFilter: (groupBy: QlceViewFieldInputInput) => QlceViewGroupByInput = groupBy => {
  return {
    entityGroupBy: groupBy
  } as QlceViewGroupByInput
}

export const getTimeRangeFilter: (aggregation: QlceViewTimeGroupType) => QlceViewGroupByInput = aggregation => {
  return {
    timeTruncGroupBy: {
      resolution: aggregation
    }
  } as QlceViewGroupByInput
}

export const getFilters: (viewConditions: QlceViewFilterInput[]) => QlceViewFilterWrapperInput[] = viewConditions => {
  return viewConditions
    .filter(viewCondition => {
      return viewCondition.values?.length
    })
    .map(viewCondition => {
      return {
        idFilter: {
          values: viewCondition.values,
          operator: viewCondition.operator,
          field: {
            fieldId: viewCondition.field.fieldId,
            fieldName: viewCondition.field.fieldName,
            identifier: viewCondition.field.identifier
          }
        }
      }
    }) as QlceViewFilterWrapperInput[]
}

export const DEFAULT_GROUP_BY = {
  fieldId: 'product',
  fieldName: 'Product',
  identifier: ViewFieldIdentifier.Common,
  identifierName: ViewFieldIdentifier.Common
}

export const GROUP_BY_POD = {
  fieldId: 'instanceId',
  fieldName: 'Pod',
  identifier: ViewFieldIdentifier.Cluster,
  identifierName: ViewFieldIdentifier.Cluster
}

export const generateId: (length: number) => string = length => {
  let result = ''
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
  const charactersLength = characters.length
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength))
  }
  return result
}

export const CREATE_CALL_OBJECT = {
  viewVersion: 'v1',
  viewTimeRange: {
    viewTimeRangeType: ViewTimeRangeType.Last_7
  },
  viewType: 'CUSTOMER',
  viewVisualization: {
    granularity: QlceViewTimeGroupType.Day,
    groupBy: { ...DEFAULT_GROUP_BY },
    chartType: ViewChartType.StackedLineChart
  }
}

export const normalizeViewRules: (viewRules: ViewRule[] | undefined) => QlceViewRuleInput[] = viewRules => {
  return viewRules
    ? (viewRules
        .map(rule => {
          if (rule) {
            return {
              conditions:
                rule.viewConditions &&
                rule.viewConditions
                  .map(c => {
                    const condition = c as unknown as ViewIdCondition
                    return condition.values && condition.values.length
                      ? {
                          field: {
                            fieldId: condition.viewField?.fieldId,
                            fieldName: condition.viewField?.fieldName,
                            identifier: condition.viewField?.identifier,
                            identifierName: condition.viewField?.identifierName
                          },
                          operator: condition.viewOperator,
                          values: condition.values
                        }
                      : null
                  })
                  .filter(condition => !!condition)
            }
          }
        })
        .filter(rule => rule?.conditions && rule.conditions.length) as QlceViewRuleInput[])
    : []
}

export const getRuleFilters: (rules: QlceViewRuleInput[]) => QlceViewFilterWrapperInput[] = rules => {
  return rules.map(rule => ({
    ruleFilter: rule
  })) as QlceViewFilterWrapperInput[]
}

export const SOURCE_ICON_MAPPING: Record<string, IconName> = {
  AWS: 'service-aws',
  GCP: 'gcp',
  CLUSTER: 'blue-black-cluster',
  CUSTOM: 'pipeline-custom',
  AZURE: 'service-azure'
}

export const perspectiveDefaultTimeRangeMapper: Record<string, moment.Moment[]> = {
  [ViewTimeRangeType.Last_7]: DATE_RANGE_SHORTCUTS.LAST_7_DAYS,
  [ViewTimeRangeType.Last_30]: DATE_RANGE_SHORTCUTS.LAST_30_DAYS,
  [ViewTimeRangeType.LastMonth]: DATE_RANGE_SHORTCUTS.LAST_MONTH
}

export enum ChartState {
  IN_ACTIVE = 'inactive',
  NONE = ''
}

export const highlightNode = (chartRef: React.RefObject<Highcharts.Chart | undefined>, id: string) => {
  if (chartRef && chartRef.current) {
    const chart = chartRef.current
    chart.series?.length > 1 &&
      chart.series.forEach(chartItem => {
        const nodeId = (chartItem.options as any).nodeId
        if (nodeId !== id) {
          chartItem.setState(ChartState.IN_ACTIVE)
        }
      })
  }
}

export const resetNodeState = (chartRef: React.RefObject<Highcharts.Chart | undefined>) => {
  if (chartRef && chartRef.current) {
    const chart = chartRef.current
    chart.series &&
      chart.series.forEach(chartItem => {
        chartItem.setState(ChartState.NONE)
      })
  }
}

export const clusterInfoUtil: (dataSources?: string[]) => { isClusterOnly: boolean; hasClusterAsSource: boolean } =
  dataSources => {
    let isClusterOnly = false
    let hasClusterAsSource = false
    if (!dataSources?.length) {
      return { isClusterOnly, hasClusterAsSource }
    }
    if (dataSources.length === 1 && dataSources[0] === ViewFieldIdentifier.Cluster) {
      isClusterOnly = true
    }
    if (dataSources.includes(ViewFieldIdentifier.Cluster)) {
      hasClusterAsSource = true
    }
    return { isClusterOnly, hasClusterAsSource }
  }
