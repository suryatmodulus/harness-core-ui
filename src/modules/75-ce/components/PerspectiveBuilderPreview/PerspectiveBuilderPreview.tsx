import React from 'react'
import { useParams } from 'react-router-dom'
import { Container, Text, Layout, FlexExpander, Icon, TextInput, FontVariation } from '@wings-software/uicore'
import cx from 'classnames'
import { Menu, MenuItem, Popover, Position } from '@blueprintjs/core'
import type moment from 'moment'
import { noop } from 'lodash-es'
import { useStrings } from 'framework/strings'
import type { CEView } from 'services/ce'
import {
  useFetchViewFieldsQuery,
  QlceViewFilterWrapperInput,
  useFetchPerspectiveFiltersValueQuery,
  QlceViewFieldInputInput,
  ViewFieldIdentifier,
  ViewChartType,
  useFetchPerspectiveTimeSeriesQuery,
  QlceViewTimeGroupType,
  useFetchperspectiveGridQuery,
  ViewTimeRangeType,
  Maybe,
  QlceViewFieldIdentifierData,
  QlceViewField
} from 'services/ce/services'
import CloudCostInsightChart from '@ce/components/CloudCostInsightChart/CloudCostInsightChart'
import {
  normalizeViewRules,
  getRuleFilters,
  getGroupByFilter,
  getTimeRangeFilter,
  getTimeFilters,
  getViewFilterForId
} from '@ce/utils/perspectiveUtils'
import { CCM_CHART_TYPES } from '@ce/constants'
import { DATE_RANGE_SHORTCUTS } from '@ce/utils/momentUtils'
import { DAYS_FOR_TICK_INTERVAL } from '@ce/components/CloudCostInsightChart/Chart'
import { AGGREGATE_FUNCTION } from '../PerspectiveGrid/Columns'
import PerspectiveGrid from '../PerspectiveGrid/PerspectiveGrid'
import css from './PerspectiveBuilderPreview.module.scss'

interface GroupByViewSubMenuProps {
  labelData: Maybe<Maybe<string>[]>
  field: QlceViewFieldIdentifierData
  setGroupBy: (groupBy: QlceViewFieldInputInput) => void
}

const GroupByViewSubMenu: (props: GroupByViewSubMenuProps) => JSX.Element | null = ({
  field,
  labelData,
  setGroupBy
}) => {
  const { getString } = useStrings()
  const [searchText, setSearchText] = React.useState('')

  const filteredLabelData = (labelData || []).filter(label => {
    if (!label) {
      return false
    }
    return label.toLocaleLowerCase().indexOf(searchText.toLocaleLowerCase()) < 0 ? false : true
  })

  const renderLabels: (value: QlceViewField) => void = value => {
    return (
      <MenuItem className={css.menuItem} key={value.fieldId} text={value.fieldName}>
        <div className={css.groupByLabel}>
          <TextInput
            value={searchText}
            onChange={(e: any) => {
              setSearchText(e.target.value)
            }}
            placeholder={getString('ce.perspectives.createPerspective.filters.searchText')}
          />
          <Container className={css.labelValueContainer}>
            {filteredLabelData.map(label => (
              <MenuItem
                className={css.menuItem}
                key={label}
                text={label}
                onClick={() =>
                  setGroupBy({
                    identifier: ViewFieldIdentifier.Label,
                    fieldId: 'labels.value',
                    fieldName: label || '',
                    identifierName: 'Label'
                  })
                }
              />
            ))}
          </Container>
        </div>
      </MenuItem>
    )
  }

  if (field.values.length) {
    return (
      <>
        {field.values.map(value => {
          if (value) {
            if (value.fieldId === 'label') {
              return renderLabels(value)
            }
            return (
              <MenuItem
                className={css.menuItem}
                key={value.fieldId}
                text={value.fieldName}
                onClick={() =>
                  setGroupBy({
                    fieldId: value.fieldId,
                    fieldName: value.fieldName,
                    identifier: field.identifier,
                    identifierName: field.identifierName
                  })
                }
              />
            )
          }
          return null
        })}
      </>
    )
  }
  return null
}

interface GroupByViewProps {
  groupBy: QlceViewFieldInputInput
  setGroupBy: (groupBy: QlceViewFieldInputInput) => void
  chartType: ViewChartType
  setChartType: (type: ViewChartType) => void
}

const GroupByView: React.FC<GroupByViewProps> = ({ groupBy, setGroupBy, chartType, setChartType }) => {
  const { perspectiveId } = useParams<{ perspectiveId: string }>()

  const { getString } = useStrings()

  const [result] = useFetchViewFieldsQuery({
    variables: {
      filters: [{ viewMetadataFilter: { viewId: perspectiveId, isPreview: true } } as QlceViewFilterWrapperInput]
    }
  })
  const { data, fetching } = result

  const [labelResult] = useFetchPerspectiveFiltersValueQuery({
    variables: {
      filters: [
        {
          idFilter: {
            field: {
              fieldId: 'labels.key',
              fieldName: '',
              identifier: 'LABEL'
            },
            operator: 'IN',
            values: []
          }
        } as unknown as QlceViewFilterWrapperInput
      ],
      offset: 0,
      limit: 100
    }
  })

  const { data: labelResData } = labelResult

  const labelData = labelResData?.perspectiveFilters?.values

  const fieldIdentifierData = data?.perspectiveFields?.fieldIdentifierData

  const PopoverContent =
    fieldIdentifierData && fieldIdentifierData.length ? (
      <Menu>
        {fieldIdentifierData.map(field => {
          if (field) {
            return (
              <MenuItem className={css.menuItem} key={field.identifier} text={field.identifierName}>
                <GroupByViewSubMenu setGroupBy={setGroupBy} labelData={labelData || []} field={field} />
              </MenuItem>
            )
          }
          return null
        })}
      </Menu>
    ) : undefined

  return (
    <Layout.Horizontal
      spacing="small"
      style={{
        alignItems: 'center'
      }}
    >
      <Text font="small" color="grey400">
        {getString('ce.perspectives.createPerspective.preview.groupBy')}
      </Text>
      <Popover
        disabled={fetching}
        position={Position.BOTTOM_LEFT}
        modifiers={{
          arrow: { enabled: false },
          flip: { enabled: true },
          keepTogether: { enabled: true },
          preventOverflow: { enabled: true }
        }}
        content={PopoverContent}
      >
        <Container background="grey100" className={cx(css.groupBySelect, { [css.groupBySelectLoading]: fetching })}>
          {fetching ? (
            <Icon name="spinner" />
          ) : (
            <Text font="small" background="grey100" rightIcon="chevron-down">
              {groupBy.fieldName}
            </Text>
          )}
        </Container>
      </Popover>
      <FlexExpander />
      <Container>{/* <Text font="small">Aggregation</Text> */}</Container>
      <Icon
        name="timeline-bar-chart"
        size={18}
        onClick={() => {
          setChartType(ViewChartType.StackedTimeSeries)
        }}
        color={chartType === ViewChartType.StackedTimeSeries ? 'primary6' : 'grey500'}
      />
      <Icon
        name="timeline-area-chart"
        size={20}
        onClick={() => {
          setChartType(ViewChartType.StackedLineChart)
        }}
        color={chartType === ViewChartType.StackedLineChart ? 'primary6' : 'grey500'}
      />
    </Layout.Horizontal>
  )
}

interface PerspectiveBuilderPreviewProps {
  groupBy: QlceViewFieldInputInput
  setGroupBy: (groupBy: QlceViewFieldInputInput) => void
  chartType: ViewChartType
  setChartType: (type: ViewChartType) => void
  formValues: CEView
}

const PerspectiveBuilderPreview: React.FC<PerspectiveBuilderPreviewProps> = ({
  groupBy,
  setGroupBy,
  chartType,
  setChartType,
  formValues
}) => {
  const { perspectiveId } = useParams<{ perspectiveId: string }>()

  const timeRangeMapper: Record<string, moment.Moment[]> = {
    [ViewTimeRangeType.Last_7]: DATE_RANGE_SHORTCUTS.LAST_7_DAYS,
    [ViewTimeRangeType.Last_30]: DATE_RANGE_SHORTCUTS.LAST_30_DAYS,
    [ViewTimeRangeType.LastMonth]: DATE_RANGE_SHORTCUTS.LAST_MONTH
  }

  const timeRange = formValues.viewTimeRange?.viewTimeRangeType || ViewTimeRangeType.Last_7

  const dateRange = timeRangeMapper[timeRange]
  const [chartResult] = useFetchPerspectiveTimeSeriesQuery({
    variables: {
      filters: [
        getViewFilterForId(perspectiveId, true),
        ...getTimeFilters(dateRange[0].valueOf(), dateRange[1].valueOf()),
        ...getRuleFilters(normalizeViewRules(formValues.viewRules))
      ],
      limit: 12,
      groupBy: [
        getTimeRangeFilter(
          (formValues.viewVisualization?.granularity as QlceViewTimeGroupType) || QlceViewTimeGroupType.Day
        ),
        getGroupByFilter(groupBy)
      ]
    }
  })

  const [gridResults] = useFetchperspectiveGridQuery({
    variables: {
      aggregateFunction: AGGREGATE_FUNCTION.DEFAULT,
      filters: [
        getViewFilterForId(perspectiveId, true),
        ...getTimeFilters(dateRange[0].valueOf(), dateRange[1].valueOf()),
        ...getRuleFilters(normalizeViewRules(formValues.viewRules))
      ],
      limit: 100,
      offset: 0,
      isClusterOnly: false,
      groupBy: [getGroupByFilter(groupBy)]
    }
  })

  const { data: gridData, fetching: gridFetching } = gridResults
  const { data: chartData, fetching } = chartResult

  const { getString } = useStrings()
  return (
    <Container padding="xxlarge" background="white" className={css.previewMainContainer}>
      <Layout.Vertical spacing="xlarge">
        <Container>
          <Text font={{ variation: FontVariation.H4 }} margin={{ bottom: 'medium' }}>
            {getString('ce.perspectives.createPerspective.preview.title')}
          </Text>
          <GroupByView setGroupBy={setGroupBy} groupBy={groupBy} chartType={chartType} setChartType={setChartType} />

          <CloudCostInsightChart
            chartType={chartType === ViewChartType.StackedLineChart ? CCM_CHART_TYPES.AREA : CCM_CHART_TYPES.COLUMN}
            columnSequence={[]}
            setFilterUsingChartClick={() => {
              noop
            }}
            fetching={fetching}
            showLegends={false}
            data={chartData?.perspectiveTimeSeriesStats as any}
            aggregation={
              (formValues.viewVisualization?.granularity as QlceViewTimeGroupType) || QlceViewTimeGroupType.Day
            }
            xAxisPointCount={chartData?.perspectiveTimeSeriesStats?.stats?.length || DAYS_FOR_TICK_INTERVAL + 1}
          />
        </Container>
        <Container width="100%">
          <Text color="grey900">Cost Breakdown</Text>
          {gridData?.perspectiveGrid?.data && (
            <PerspectiveGrid
              gridFetching={gridFetching}
              gridData={gridData?.perspectiveGrid?.data as any}
              groupBy={groupBy}
              showColumnSelector={false}
              tempGridColumns={true}
              showPagination={false}
            />
          )}
        </Container>
      </Layout.Vertical>
    </Container>
  )
}

export default PerspectiveBuilderPreview
