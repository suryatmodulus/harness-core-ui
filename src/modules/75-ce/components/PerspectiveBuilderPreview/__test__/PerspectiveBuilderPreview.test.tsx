/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'urql'
import { fromValue } from 'wonka'
import type { DocumentNode } from 'graphql'
import { TestWrapper } from '@common/utils/testUtils'
import {
  FetchPerspectiveTimeSeriesDocument,
  FetchPerspectiveDetailsSummaryDocument,
  FetchViewFieldsDocument,
  ViewChartType,
  FetchperspectiveGridDocument
} from 'services/ce/services'

import ChartResponseData from '@ce/pages/perspective-details/__test__/ChartDataResponse.json'
import SummaryResponseData from '@ce/pages/perspective-details/__test__/SummaryResponse.json'
import ViewFieldResponseData from '@ce/pages/perspective-details/__test__/ViewFieldResponse.json'
import { DEFAULT_GROUP_BY } from '@ce/utils/perspectiveUtils'
import PerspectiveBuilderPreview from '../PerspectiveBuilderPreview'

jest.mock('@ce/components/CEChart/CEChart', () => 'mock')

const params = {
  accountId: 'TEST_ACC',
  perspetiveId: 'perspectiveId',
  perspectiveName: 'sample perspective'
}

const props = {
  groupBy: DEFAULT_GROUP_BY,
  setGroupBy: jest.fn(),
  chartType: ViewChartType.StackedLineChart,
  setChartType: jest.fn,
  formValues: {}
}

describe('test cases for Perspective details Page', () => {
  test('should be able to render the details page', async () => {
    const responseState = {
      executeQuery: ({ query }: { query: DocumentNode }) => {
        if (query === FetchPerspectiveTimeSeriesDocument) {
          return fromValue(ChartResponseData)
        }
        if (query === FetchperspectiveGridDocument) {
          return fromValue({})
        }
        if (query === FetchPerspectiveDetailsSummaryDocument) {
          return fromValue(SummaryResponseData)
        }
        if (query === FetchViewFieldsDocument) {
          return fromValue(ViewFieldResponseData)
        }
        return fromValue({})
      }
    }

    const { container } = render(
      <TestWrapper pathParams={params}>
        <Provider value={responseState as any}>
          <PerspectiveBuilderPreview {...props} />
        </Provider>
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
