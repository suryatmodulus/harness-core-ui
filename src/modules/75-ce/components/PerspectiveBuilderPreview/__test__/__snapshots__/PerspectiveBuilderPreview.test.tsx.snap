// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test cases for Perspective details Page should be able to render the details page 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main previewMainContainer StyledProps--padding-xxlarge StyledProps--background-white"
  >
    <div
      class="Layout--vertical Layout--layout-spacing-xlarge StyledProps--main"
    >
      <div
        class="StyledProps--font StyledProps--main"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-h4 StyledProps--margin-bottom-medium"
        >
          ce.perspectives.createPerspective.preview.title
        </p>
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main"
          style="align-items: center;"
        >
          <p
            class="StyledProps--font StyledProps--main StyledProps--font-small StyledProps--color StyledProps--color-grey400"
          >
            ce.perspectives.createPerspective.preview.groupBy
          </p>
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <div
                class="StyledProps--font StyledProps--main groupBySelect StyledProps--background-grey100"
              >
                <p
                  class="StyledProps--font StyledProps--main StyledProps--font-small StyledProps--background-grey100"
                  style="display: flex; align-items: center;"
                >
                  Product
                  <span
                    class="bp3-icon bp3-icon-chevron-down StyledProps--main Text--icon StyledProps--padding-left-xsmall"
                    icon="chevron-down"
                  >
                    <svg
                      data-icon="chevron-down"
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                    >
                      <desc>
                        chevron-down
                      </desc>
                      <path
                        d="M12 5c-.28 0-.53.11-.71.29L8 8.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42l4 4c.18.18.43.29.71.29s.53-.11.71-.29l4-4A1.003 1.003 0 0012 5z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </p>
              </div>
            </span>
          </span>
          <span
            style="flex-grow: 1;"
          />
          <div
            class="StyledProps--font StyledProps--main"
          />
          <span
            class="bp3-icon bp3-icon-timeline-bar-chart StyledProps--main StyledProps--color StyledProps--color-grey500"
            icon="timeline-bar-chart"
          >
            <svg
              data-icon="timeline-bar-chart"
              height="18"
              viewBox="0 0 16 16"
              width="18"
            >
              <desc>
                timeline-bar-chart
              </desc>
              <path
                d="M8 12h1c.55 0 1-.45 1-1V2c0-.55-.45-1-1-1H8c-.55 0-1 .45-1 1v9c0 .55.45 1 1 1zm5 0h1c.55 0 1-.45 1-1V5c0-.55-.45-1-1-1h-1c-.55 0-1 .45-1 1v6c0 .55.45 1 1 1zm2 1H2c-.55 0-1 .45-1 1s.45 1 1 1h13c.55 0 1-.45 1-1s-.45-1-1-1zM3 12h1c.55 0 1-.45 1-1V7c0-.55-.45-1-1-1H3c-.55 0-1 .45-1 1v4c0 .55.45 1 1 1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-icon bp3-icon-timeline-area-chart StyledProps--main StyledProps--color StyledProps--color-primary6"
            icon="timeline-area-chart"
          >
            <svg
              data-icon="timeline-area-chart"
              height="20"
              viewBox="0 0 20 20"
              width="20"
            >
              <desc>
                timeline-area-chart
              </desc>
              <path
                d="M19 16H2V3c0-.55-.45-1-1-1s-1 .45-1 1v14c0 .55.45 1 1 1h18c.55 0 1-.45 1-1s-.45-1-1-1zm0-13.41l-7.07 7.07-4.3-3.44-.01.01A.987.987 0 007 6c-.24 0-.46.1-.63.24l-.01-.01L3 9.03V15h16V2.59z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
        <div>
          <article>
            <mock
              options="[object Object]"
            />
          </article>
        </div>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
        style="width: 100%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-grey900"
        >
          Cost Breakdown
        </p>
      </div>
    </div>
  </div>
</div>
`;
